tensorflow/python/ops/math_ops_test.py

```diff
@@ -44,6 +44,16 @@ def testReduceAllDims(self):
       y_tf = self.evaluate(math_ops.reduce_sum(x))
       self.assertEqual(y_tf, 21)
 
+  def testReduceExtendType(self):
+    in_f32 = np.random.randn(1000, 1000).astype(np.float32)
+    in_bf16 = math_ops.cast(in_f32, dtypes.bfloat16)
+
+    out_f32 = self.evaluate(math_ops.reduce_sum(in_f32))
+    out_bf16 = self.evaluate(math_ops.reduce_sum(in_bf16))
+    expected = math_ops.cast(out_f32, dtypes.bfloat16)
+
+    self.assertAllEqual(out_bf16, expected)
```

Comment:
If the reduction is run in parallel (i.e. ThreadPoolDevice), the summation orders in both reductions could be different, causing the results to not match exactly (because of round-off errors). Could you please use `assertAllClose` instead, and pick a tolerance/threshold where the answer would match only if `f32` was used in the computation?

Reply:
Thanks for the suggestion, before modify the code I have a simple question: in original code  ThreadPoolDevice is called [here](https://github.com/tensorflow/tensorflow/pull/38630/files#diff-dce20cd1a37fce5a056d1f1c1fa02c08R75) after accumulate type conversion, so I don't think it will be rounded to low precision value in parallel. Does TF have a way to use nested ThreadPoolDevice? Or just make safe in case user use this way?

Reply:
@penpornk Can you please take a look on the above comment from @Zantares. Thanks!

Reply:
Thank you for the kindly reminding,  I ask this question is because we want to ensure accumulate type computation won't be broken in TF process, otherwise BF16 will lose accuracy. This is also the reason I use `assertAllClose` for BF16 Softmax and SoftmaxGrad  meanwhile `assertAllEqual` for `SumReduce`, we don't expect any accuracy loss happening in an operation with accumulate type inside only.

Reply:
Sorry for the delay! Somehow I missed @Zantares's reply the first time. Thank you for the reminder and the clarification.

My problem is with the rounding errors **within** the type f32 due to parallel computation (I will explain how it's parallelized in the next paragraph). In theory, `out_f32` in this test should have exactly the same results as the intermediate f32 results for `out_bf16` (I'll call this `out_bf16_f32_accumulator`). But since the reduction happen in parallel, there can be different roundoff errors due to different summation orders. For example, `out_f32` could be 0.9980468 and `out_bf16_f32_accumulator` could be 0.9980469, which will be 0.99609375 and 1 when cast to bfloat16, so we shouldn't use `assertAllEqual`.

Back to this reduce expression (similar to [yours](https://github.com/tensorflow/tensorflow/pull/38630/files#diff-dce20cd1a37fce5a056d1f1c1fa02c08R75) -- I just don't know how to quote a line in a PR):
https://github.com/tensorflow/tensorflow/blob/37543935526ec5c28893e448c5ee29f24b9d2aee/tensorflow/core/kernels/reduction_ops.h#L57

Eigen evaluates expressions lazily, i.e., accumulates operations through nested template typing, and only compute all operations when the expression is assigned to a device by calling the device's `TensorExecutor`. The `TensorExecutor` recursively evaluates the expression. If the device is a `ThreadPoolDevice`, evaluators can split work into shards and schedule them onto the thread pool.

`SumReducer` has an evaluator that eventually calls [Eigen::internal::FullReducer](https://gitlab.com/libeigen/eigen/-/blob/8a7f360ec3160acdb545d05712435d8ca0334357/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h#L323) that [shards the work and compute in parallel](https://gitlab.com/libeigen/eigen/-/blob/8a7f360ec3160acdb545d05712435d8ca0334357/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h#L357). 

An example call stack (most recent call on the top):
```
Eigen::internal::SumReducer<>::reduce()
Eigen::internal::InnerMostDimReducer<>::reduce()
Eigen::internal::FullReducer<>::run()
Eigen::TensorReductionEvaluatorBase<>::evalSubExprsIfNeededCommon()
Eigen::TensorReductionEvaluatorBase<>::evalSubExprsIfNeeded()
Eigen::TensorEvaluator<>::evalSubExprsIfNeeded()
Eigen::internal::TensorExecutor<>::run()
Eigen::TensorDevice<>::operator=<>()
tensorflow::functor::ReduceEigenImpl<>::operator()()
tensorflow::functor::ReduceFunctorBase<>::Reduce<>()
tensorflow::ReductionOp<>::Compute()
```

Reply:
Got it, so you mean the rounding error is happening inside fp32 computation which is parallelized by Eigen reducer, the accumulate type output `out_bf16_f32_accumulator` could be different with multi-running. Thanks again for the experience sharing and I will address the code soon.

Reply:
Done, replace `assertAllEqual` to `assertAllClose` with a generic tolerance `1e-3` .